MekHQ/src/mekhq/service/AutosaveService.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * AutosaveService.java
+ *
+ * Copyright (c) 2019 MekHQ Team. All rights reserved.
+ *
+ * This file is part of MekHQ.
+ *
+ * MekHQ is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * MekHQ is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MekHQ.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package mekhq.service;
+
+import megamek.common.logging.MMLogger;
+import mekhq.MekHQ;
+import mekhq.MekHqConstants;
+import mekhq.campaign.Campaign;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.prefs.Preferences;
+import java.util.stream.Collectors;
+import java.util.zip.GZIPOutputStream;
+
+public class AutosaveService implements IAutosaveService {
+    private final Preferences userPreferences = Preferences.userRoot().node(MekHqConstants.AUTOSAVE_NODE);
+    private final MMLogger logger;
+
+    public AutosaveService(MMLogger logger) {
+        assert logger != null;
+
+        this.logger = logger;
+    }
+
+    @Override
+    public void requestDayAdvanceAutosave(Campaign campaign, int dayOfTheWeek) {
+        assert campaign != null;
+
+        if (this.isDailyAutosaveEnabled()) {
+            this.performAutosave(campaign);
+        } else if (dayOfTheWeek == Calendar.SUNDAY && this.isWeeklyAutosaveEnabled()) {
+            this.performAutosave(campaign);
+        }
+    }
+
+    public void requestBeforeMissionAutosave(Campaign campaign) {
+        assert campaign != null;
+
+        if (this.isMissionAutosaveEnabled()) {
+            this.performAutosave(campaign);
+        }
+    }
+
+    private boolean isDailyAutosaveEnabled() {
+        return this.userPreferences.getBoolean(MekHqConstants.SAVE_DAILY_KEY, false);
+    }
+
+    private boolean isWeeklyAutosaveEnabled() {
+        return this.userPreferences.getBoolean(MekHqConstants.SAVE_WEEKLY_KEY, false);
+    }
+
+    private boolean isMissionAutosaveEnabled() {
+        return this.userPreferences.getBoolean(MekHqConstants.SAVE_BEFORE_MISSIONS_KEY, false);
+    }
+
+    private void performAutosave(Campaign campaign) {
+        try {
+            String fileName = this.getAutosaveFilename(campaign);
+
+            try (GZIPOutputStream output = new GZIPOutputStream(new FileOutputStream(fileName))) {
```

Comment:
I believe this will fix the LGTM alert:
```suggestion
            try (FileOutputStream os = new FileOutputStream(fileName);
                 GZIPOutputStream output = new GZIPOutputStream(os)) {
```
Based on https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

Reply:
Interesting, I thought that closing the outer stream will automatically close the inner one (like in .NET). Thanks for the info!